### PR TITLE
[ci] Add rspec test for AttributeController#create

### PR DESCRIPTION
**Pull request by eduardoj**

### CODE REVIEW

**Comment by eduardoj** [comment]
[(1) P6: app/controllers/webui/attribute_controller.rb](https://trello.com/c/POBX2dU3/300-1-p6-app-controllers-webui-attribute-controller-rb)

**Comment by hennevogel** [comment]
What about all the other conditions?

**Comment by eduardoj** [comment]
@hennevogel , I didn't understand correctly. I will test all the other conditions.

By the way, are you referring to the `create` action only, or to all the other conditions of all the actions of this controller?

File: src/api/spec/controllers/webui/attributes_spec.rb [hunk 1]
```diff
@@ -16,4 +16,20 @@
...
+
+  describe 'POST #create' do
+    let(:attribute_type) { create(:attrib_type_with_namespace, value_count: 1) }
+
+    before do
+      login user
+      post :create, params: { attrib: { project_id: user.home_project.id, attrib_type_id: attribute_type.id } }
+    end
+
+    it {
```
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: Do not use `{...}` for multi-line blocks: https://github.com/bbatsov/ruby-style-guide#single-line-blocks

Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: WE should enable the Rubocop cop for this :thinking:

File: src/api/spec/controllers/webui/attributes_spec.rb [hunk 2]
```diff
@@ -16,4 +16,20 @@
...
+      expect(response).to redirect_to(edit_attribs_path(project: user.home_project_name,
+                                                        package: '',
```
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: do you need to specify the `package`?

Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: If it is not specified this is the error:

```
Failures:

  1) Webui::AttributeController POST #create with value_count = 1 should redirect to "/attribs/home:user_3/ut:attribute_factory_3/edit"
     Failure/Error:
       expect(response).to redirect_to(edit_attribs_path(project: user.home_project_name,
                                                         attribute: "#{attribute_type.namespace}:#{attribute_type.name}"))
     
       Expected response to be a redirect to <http://test.host/attribs/home:user_3/ut:attribute_factory_3/edit> but was a redirect to <http://test.host/attribs/home:user_3//ut:attribute_factory_3/edit>.
       Expected "http://test.host/attribs/home:user_3/ut:attribute_factory_3/edit" to be === "http://test.host/attribs/home:user_3//ut:attribute_factory_3/edit".
     # ./spec/controllers/webui/attributes_spec.rb:30:in `block (4 levels) in <top (required)>'
     # ./spec/support/logging.rb:4:in `block (2 levels) in <top (required)>'
```

Can we use other option?

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 1 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: @Ana06 Done. Please re-review.

**Comment by bgeuken** [comment]
LGTM

File: src/api/spec/controllers/webui/attributes_spec.rb [hunk 3]
```diff
@@ -16,4 +16,42 @@
...
+    let(:attribute_type_1) { create(:attrib_type_with_namespace, value_count: 1) }
+    let(:attribute_type_1_name) { "#{attribute_type_1.namespace}:#{attribute_type_1.name}" }
+
+    before do
+      login user
+    end
+
+    context 'with editable values' do
+      before do
+        post :create, params: { attrib: { project_id: user.home_project.id, attrib_type_id: attribute_type_1.id } }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: you shouldn't need the id:

`project: user.home_project, attrib_type: attribute_type_1`

File: src/api/spec/controllers/webui/attributes_spec.rb [hunk 4]
```diff
@@ -16,4 +16,42 @@
...
+      end
+
+      it { expect(response).to redirect_to(edit_attribs_path(project: user.home_project_name, package: '', attribute: attribute_type_1_name)) }
+      it { expect(flash[:notice]).to eq('Attribute was successfully created.') }
+    end
+
+    context 'with non editable values' do
+      before do
+        post :create, params: { attrib: { project_id: user.home_project.id, attrib_type_id: attribute_type_0.id } }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: same here

File: src/api/spec/controllers/webui/attributes_spec.rb [hunk 5]
```diff
@@ -16,4 +16,42 @@
...
+    end
+
+    context 'fails at save' do
+      before do
+        allow_any_instance_of(Attrib).to receive(:save).and_return(false)
+        post :create, params: { attrib: { project_id: user.home_project.id, attrib_type_id: attribute_type_1.id } }
+      end
+
+      it { expect(response).to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq('Saving attribute failed: ') }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: The problem of how you mock the save is that you generate a non real error message.

The code looks like:

```Ruby
redirect_back(fallback_location: root_path, error: "Saving attribute failed: #{@attribute.errors.full_messages.join(', ')}")
```

And as you mocked the `save` there are not `errors` for the attribute. So then this is not a real case, and I would really prefer testing things that can actually happen.

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 2 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: I meant to use `nil`, but doesn't make a difference and I am not sure if It would work. It is perfectly fine as it is. :wink:

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 3 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: It seems that it is needed:

```
Failures:

  1) Webui::AttributeController POST #create with editable values
     Failure/Error:
       params.require(:attrib).permit(:attrib_type_id, :project_id, :package_id,
                                      values_attributes: [:id, :value, :position, :_destroy],
                                      issues_attributes: [:id, :name, :issue_tracker_id, :_destroy])

     ActionController::UnpermittedParameters:
       found unpermitted parameters: attrib_type, project
     # ./app/controllers/webui/attribute_controller.rb:113:in `attrib_params'
     # ./app/controllers/webui/attribute_controller.rb:32:in `create'
     # ./spec/controllers/webui/attributes_spec.rb:31:in `block (4 levels) in <top (required)>'
     # ./spec/support/logging.rb:4:in `block (2 levels) in <top (required)>'
```

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 4 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: Same error.

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 5 above)
Comment (mdeniz) [inline src/api/spec/controllers/webui/attributes_spec.rb]: I don't like to test the error message neither... so, for me is ok to just to mock the save and do something like this:
```ruby
 it { expect(flash[:error]).not_to be_nil }
```

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 3 above)
Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: Why are we using `project_id` instead of `project` in the strong params? :rage: 

ok, we could change it but this is unrelated to the PR, so leave it as it is.

File: src/api/spec/controllers/webui/attributes_spec.rb (hunk 5 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: @Ana06 In this case, 'things that can actually happen' to make this `save` call return `false` would be better tested in a model or feature test, I think.

Can we agree on that?

Comment (Ana06) [inline src/api/spec/controllers/webui/attributes_spec.rb]: > @Ana06 In this case, 'things that can actually happen' to make this save call return false would be better tested in a model or feature test, I think.
> Can we agree on that?

no, feature test are for testing workflows and model test for model methods. Nothing to do with testing real cases or not.

> I don't like to test the error message neither... so, for me is ok to just to mock the save and do something like this:

I would prefer to generate a possible error case and test the error message. But as the discussion about testing error messages has already came up in several PR, I am also fine with @mdeniz suggestion.

Comment (eduardoj) [inline src/api/spec/controllers/webui/attributes_spec.rb]: @Ana06 @mdeniz Done.
